sw/device/lib/dif/dif_hmac.h

```diff
@@ -0,0 +1,235 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 1.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_DIF_HMAC_DEVICE_H_
+#define SW_DEVICE_LIB_DIF_HMAC_DEVICE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Supported HMAC modes of operation.
+ */
+typedef enum dif_hmac_mode {
+  /** The HMAC mode */
+  kDifHmacModeHmac = 0,
+  /** The SHA256-only mode */
+  kDifHmacModeSha256,
+} dif_hmac_mode_t;
+
+/**
+ * Supported byte endienness options.
+ */
+typedef enum dif_hmac_endianness {
+  /** Big endian byte ordering */
+  kDifHmacEndiannessBig = 0,
+  /** Little endian byte ordering */
+  kDifHmacEndiannessLittle,
+} dif_hmac_endianness_t;
+
+/**
+ * Error codes for HMAC functions that may exhibit generic failures.
+ */
+typedef enum dif_hmac_error {
+  /** No error occurred. */
+  kDifHmacErrorNone = 0,
+  /** An invalid argument was provided. */
+  kDifHmacErrorInvalidArgument,
+} dif_hmac_error_t;
+
+/**
+ * Error codes for HMAC FIFO operations that may fail.
+ */
+typedef enum dif_hmac_fifo_error {
+  /** No error occurred. */
+  kDifHmacFifoErrorNone = 0,
+  /** An invalid argument was provided. */
+  kDifHmacFifoErrorInvalidArgument,
+  /** The FIFO filled up before the buffer was fully consumed. */
+  kDifHmacFifoErrorFull,
+  /** The HMAC operation was not started with |dif_hmac_start()|. */
+  kDifHmacFifoErrorInvalidState,
+} dif_hmac_fifo_error_t;
+
+/**
+ * Configuration for initializing the HMAC device.
+ */
+typedef struct dif_hmac_config {
+  /** The base address for registers in the HMAC IP*/
+  mmio_region_t base_addr;
+  /** The mode of operation */
+  dif_hmac_mode_t mode;
+  /** Byte endianness for writes to the FIFO */
+  dif_hmac_endianness_t message_endianness;
+  /** Byte endianness for reads from the digest */
+  dif_hmac_endianness_t digest_endianness;
+  /** The HMAC key. Only used in |kDifHmacModeHmac| mode. Must point to 256 bits
+   * of contiguous, allocated memory */
+  uint8_t *key;
+} dif_hmac_config_t;
+
+/**
+ * A typed representation of the HMAC digest.
+ */
+typedef struct dif_hmac_digest { uint32_t digest[8]; } dif_hmac_digest_t;
+
+/**
+ * State for a particular HMAC device.
+ *
+ * its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_hmac {
+  mmio_region_t base_addr;
+  bool hmac_started;
```

Comment:
Hey @lenary, thanks for your detailed comments. I've attempted to respond to your comments bellow. This really gets into some fundamental question about DIF design in general, so if others have thoughts on these items, I would love to hear them.

> After some reflection, I'm not sure we've found exactly the right abstraction level with this DIF yet. I have tried to provide concrete feedback below, but the overarching concern is that this DIF is operating at a level that's too high. I would like to see the abstraction level brought back down to the level of exposing hardware functionality without too much interference.
> 
> At some point we will be writing a higher-level driver which encapsulate some of the higher-level interfaces we aim for software to use when using the hardware module, and this is where things like state machine management should happen, not in the DIF.

I've also been struggling with finding the right abstraction level for this DIF. Is `dif_hmac_fifo_push()` too high level, for example? Maybe the looping logic should live in the higher level driver and the DIF should just expose 32-bit and 8-bit writes. A function to copy a buffer into a FIFO doesn't feel too high level, but maybe it is.

It's difficult for me to reason about where the DIF abstraction should stop. If the goal is to prevent the consumer of the DIF from misusing the hardware then the DIFs need to do some amount of heavy lifting. Just exposing hardware functionality doesn't address this. This is the most difficult property of DIFs for me to understand since DIFs need to be sophisticated enough to prevent misuse of the hardware, but simple enough that they aren't replacing high level driver functionality.

> This field is approximating a state machine, but is not fully representing all the possible states, given the state machine has to cope with the loading or trashing of the secret key, and the start/complete/retrieved digest states of a single hash operation.
> 
> There are two solutions here: remove the field, or track the state machine fully. I would suggest we remove the field, as tracking the full state machine seems like a driver issue, not a DIF issue.

I'm specifically tracking this bit because `dif_hmac_fifo_push()` can lock up if HMAC isn't running. Other functions, such as `dif_hmac_digest_read()` can produce incorrect results if called at the wrong time, but something about a DIF locking up feels like a problem in a totally different category. This is sort of in line with the philosophy behind argument checking; we check some arguments for reasonable values (e.g. nullness) but not others (e.g. enum ranges). I see checking for valid state in a similar light. Both checks are catching common cases where a function's contract has been violated. We've made the decision to check for certain kinds of invalid arguments but not others, so I thought it may also be reasonable to check for certain kinds of invalid state, especially when it can avoid deadlocks.

> I had a chat with @gkelly and, upon reflection, we're not sure having more than one dif_hmac_t for the same MMIO region is particularly safe (it opens up too many concurrency issues).
> 
> This almost certainly suggests, as @silvestrst says, that there should be a way, in this DIF, to:
> 
> choose a different HMAC mode, endianness configuration,
> enable disable both the HMAC datapath and the SHA datapath
> load in a specific secret key
> These should be different from initialisation, and preferably different from each other.

I agree that multiple concurrent dif_hmac_t objects is recipe for disaster.

The `dif_hmac_init()` function is a bit of a catch-all. I could break these things out into separate functions, but this would probably be a replacement for `dif_hmac_config`, which already has all of these options.

Is there a reason to support disabling the SHA datapath? HMAC mode enables both.

Reply:
My personal theory on the abstraction boundary is thus.

Deivce X is basically a FIFO with some kind of property.
- The DIF *should* provide a "try to copy as much as possible to this FIFO" function.
- The DIF *should not* provide multiplexing and scheduling for that FIFO: that's the operating system's job.

I think that the DIF should maintain whatever state the programmer's guide for the device spec states, and no more. The view the DIF provides is as a single, global FIFO, which the operating system can safely consume without knowing the underlying details, other than that it is a FIFO that does things to the data. The view the operating system provides is the "FIFO as a service": all of the processes get to pretend they have their own FIFO, perhaps with some kind of scheduling or eviction constraints that make it possible for the operating system to multiplex the resource.